### ISSUE TITLE
NPM dedupe does not always dedupe

### ISSUE THREAD

**Issue by domoritz** [npm/npm#19007]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

### supporting information:

 - `npm -v` prints: 5.5.1
 - `node -v` prints: v8.8.1
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: MacOS

### Reproduce

```bash
mkdir temp
cd temp
npm init
npm install --save vega-embed@3.0.0-rc6 vega@2.6.1
npm ls vega-dataflow
```

actual output

```
temp@1.0.0 /private/tmp/temp
├─┬ vega@2.6.1
│ └── vega-dataflow@1.4.3
└─┬ vega-embed@3.0.0-rc6
  └─┬ vega@3.0.7
    ├─┬ vega-crossfilter@2.0.0
    │ └── vega-dataflow@3.0.3
    ├── vega-dataflow@3.0.3
    ├─┬ vega-encode@2.0.5
    │ └── vega-dataflow@3.0.3
    ├─┬ vega-force@2.0.0
    │ └── vega-dataflow@3.0.3
    ├─┬ vega-geo@2.1.1
    │ └── vega-dataflow@3.0.3
    ├─┬ vega-hierarchy@2.1.0
    │ └── vega-dataflow@3.0.3
    ├─┬ vega-parser@2.3.1
    │ └── vega-dataflow@3.0.3
    ├─┬ vega-runtime@2.0.1
    │ └── vega-dataflow@3.0.3
    ├─┬ vega-transforms@1.1.2
    │ └── vega-dataflow@3.0.3
    ├─┬ vega-view@2.0.3
    │ └── vega-dataflow@3.0.3
    ├─┬ vega-view-transforms@1.1.0
    │ └── vega-dataflow@3.0.3
    ├─┬ vega-voronoi@2.0.0
    │ └── vega-dataflow@3.0.3
    └─┬ vega-wordcloud@2.0.2
      └── vega-dataflow@3.0.3
```

`vega-dataflow` should be deduped. 

`npm dedupe` does not change anything.

**Comment by evan-scott-zocdoc** [comment]
@domoritz do you have a repository that can be looked at? Often when things don't dedupe it's due to a package being pinned down to a specific version `1.0.0` vs using a semver range like `~1.0.0` or `^1.0.0`.

**Comment by domoritz** [comment]
@evan-scott-zocdoc You can just try `npm install --save vega-embed@3.0.0-rc6 vega@2.6.1` and the look at the `package.json`. 

In the meantime, we temporarily switched to yarn, which dedupes as we expected. I still would love this to be fixed but wanted to let you know that we saw different results with yarn.

**Comment by evan-scott-zocdoc** [comment]
Yeah it's due to `vega` not using semver ranges for most of their packages: https://github.com/vega/vega/blob/master/package.json#L63-L85
https://github.com/vega/vega-wordcloud/blob/master/package.json#L30

yarn might make some assumptions NPM doesn't because NPM adheres strictly to semver 🤷‍♂️ 

vega should update to use a semver caret though

**Comment by domoritz** [comment]
@evan-scott-zocdoc, according to https://docs.npmjs.com/getting-started/semantic-versioning `3` is equivalent to `3.x` and `^3.0.0`. 

And in more detail, I found: https://docs.npmjs.com/misc/semver#advanced-range-syntax. It says "A partial version range is treated as an X-Range, so the special character is in fact optional."
